Fix crash with null callee

snichols opened the pull request:
<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->
### Issue
This PR fixes issue https://github.com/godotengine/godot/issues/40942.

### Summary
Malformed call expressions that create `GDScript::CallNode` instances with `callee == nullptr` are possible.  An example of such a malformed expression is `var crash = ()(1)`.  When these cases happen during editing, the editor crashes.

The PR fixes the bug by:
1.  Adding an accessor `GDScript::CallNode::get_callee_type` which checks for `nullptr` and returns `Type::NONE` in that case.
1.  Refactoring the code, where sensible, to use the new accessor.
1.  Adding manual `nullptr` checks where using the accessor didn't make sense.

vnen commented:
This should be solved just by adding null checks where relevant and avoiding the whole block. For instance if the `callee` is `nullptr` in the `reduce_call()` function it should just bail out early. It should not keep trying to find the type of a callee that isn't there. Same applies to other checks of the callee type.

This only happens when the parse tree is ill-formed, which happens when the code is not syntactically correct (e.g. when doing code completion on partially written code), so there's no harm in bailing out early.

snichols commented:
@vnen I've cleaned up the code to simplify and minimize the use of `get_callee_type`.

As for `reduce_call` specifically, I considered short-circuiting the function with something like:
```
if(p_call->callee == nullptr) {
	// Invalid call. Error already sent in parser.
	// TODO: Could check if Callable here too.
	p_call->set_datatype(GDScriptParser::DataType::UNRESOLVED);
	mark_node_unsafe(p_call);
	return;
}
```

I rejected the idea due to the complexity of the function and potential unwanted side effects of the change.  I'm new to this codebase and don't know the subtleties of the systems involved.

I also added better error reporting in `GDScriptParser::parse_grouping`.  The parser was reporting `Expected closing ")" after grouping expression.` when an empty grouping expression was encountered.  Now it reports `Expected grouping expression.` which seems more clear to me.

snichols commented:
Yikes.  I'll clean up this git history.  Boo for bringing in master in an unclean way!
